vcd/metadata_openapi.go

```diff
@@ -0,0 +1,329 @@
+		if err != nil {
+			return fmt.Errorf("error updating metadata with namespace '%s' and key '%s': %s", namespace, key, err)
+		}
+	}
+
+	for _, metadataEntry := range metadataToAdd {
+		_, err := resource.AddMetadata(metadataEntry)
+		if err != nil {
+			return fmt.Errorf("error adding metadata entry: %s", err)
+		}
+	}
+	return nil
+}
+
+// getOpenApiMetadataOperations retrieves the metadata that needs to be added, to be updated and to be deleted depending
+// on the old and new attribute values from Terraform state.
+func getOpenApiMetadataOperations(oldMetadata []interface{}, newMetadata []interface{}) ([]types.OpenApiMetadataEntry, map[string]types.OpenApiMetadataEntry, []string, error) {
+	oldMetadataEntries, err := getOpenApiMetadataEntryMap(oldMetadata)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	newMetadataEntries, err := getOpenApiMetadataEntryMap(newMetadata)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var metadataToRemove []string
+	for oldNamespacedKey := range oldMetadataEntries {
+		if _, ok := newMetadataEntries[oldNamespacedKey]; !ok {
+			metadataToRemove = append(metadataToRemove, oldNamespacedKey)
+		}
+	}
+
+	metadataToUpdate := map[string]types.OpenApiMetadataEntry{}
+	for newNamespacedKey, newEntry := range newMetadataEntries {
+		if oldEntry, ok := oldMetadataEntries[newNamespacedKey]; ok {
+			if reflect.DeepEqual(oldEntry, newEntry) {
+				continue
+			}
+			metadataToUpdate[newNamespacedKey] = newEntry
+		}
+	}
+
+	var metadataToCreate []types.OpenApiMetadataEntry
+	for newNamespacedKey, newEntry := range newMetadataEntries {
+		_, alreadyExisting := oldMetadataEntries[newNamespacedKey]
+		_, beingUpdated := metadataToUpdate[newNamespacedKey]
+		if !alreadyExisting && !beingUpdated {
+			metadataToCreate = append(metadataToCreate, newEntry)
+		}
+	}
+
+	return metadataToCreate, metadataToUpdate, metadataToRemove, nil
+}
+
+// getOpenApiMetadataEntryMap converts the input metadata attribute from Terraform state to a map composed by metadata
+// namespaced keys (this is, namespace and key separated by '%%%') and their values.
+func getOpenApiMetadataEntryMap(metadataAttribute []interface{}) (map[string]types.OpenApiMetadataEntry, error) {
+	metadataMap := map[string]types.OpenApiMetadataEntry{}
+	for _, rawItem := range metadataAttribute {
+		metadataEntry := rawItem.(map[string]interface{})
+
+		namespace := ""
+		if _, ok := metadataEntry["namespace"]; ok {
+			namespace = metadataEntry["namespace"].(string)
+		}
+
+		value, err := convertOpenApiMetadataValue(metadataEntry["type"].(string), metadataEntry["value"].(string))
+		if err != nil {
+			return nil, fmt.Errorf("error parsing the 'value' attribute '%s' from state: %s", metadataEntry["value"].(string), err)
+		}
+
+		// In OpenAPI, metadata is namespaced, hence it is possible to have same keys but in different namespaces
+		namespacedKey := fmt.Sprintf("%s%%%%%%%s", namespace, metadataEntry["key"].(string))
```

Comment:
IMO would be good to put a comment line about the separator string (even the detail that each `%` needs to be escaped with `%`) as I was double-checking in [playground](https://go.dev/play/p/G5D3Y0pwBcB) if it really matches https://github.com/vmware/terraform-provider-vcd/pull/1018/files#diff-a2bd1e75e42abfead79ed7048b71ae16920bc8df863d39a4e15ce5f1842e456cR151

Reply:
Done, let me know whether it is better now or is still confusing.

Reply:
Actually this suggestion made me think whether this was needed at all or I was overcomplicating things (from some leftovers of previous ideas/issues). Turns out that yes, this was unnecessary at all. So:
- This key generation is simplified, no `%%%` anymore
- The function to retrieve keys is removed.